src/client/sandbox/xhr.js

```diff
@@ -0,0 +1,173 @@
+import SandboxBase from './base';
+import nativeMethods from './native-methods';
+import { getProxyUrl } from '../utils/url';
+import XHR_HEADERS from '../../request-pipeline/xhr/headers';
+import { getOrigin } from '../utils/destination-location';
+
+const NATIVE_BEHAVIOR = 'hammerhead|xhr|native-behavior';
+
+// NOTE: We should avoid using native object prototype methods,
+// since they can be overriden by the client code. (GH-245)
+var arraySlice = Array.prototype.slice;
+
+export default class XhrSandbox extends SandboxBase {
+    constructor (sandbox) {
+        super(sandbox);
+
+        this.XHR_COMPLETED_EVENT = 'hammerhead|event|xhr-completed';
+        this.XHR_ERROR_EVENT     = 'hammerhead|event|xhr-error';
+        this.XHR_SEND_EVENT      = 'hammerhead|event|xhr-send';
+
+        var xhr = new nativeMethods.XMLHttpRequest();
+
+        this.corsSupported = typeof xhr.withCredentials !== 'undefined';
+    }
+
+    static createNativeXHR () {
+        var xhr = new window.XMLHttpRequest();

```

Comment:
@churkin please clarify the following. The method is called `createNativeXHR`. We create `window.XMLHttpRequest` that can be overriden by some script. Should not we use `nativeMethods.XMLHttpRequest` here?